File: swift/Tests/FundamentalCoreTests/GoldenConformanceTests.swift
import Foundation
import Testing
#if canImport(simd)
import simd
#endif
@testable import FundamentalCore

// Cross-plane conformance (#526) — the Swift engine must reproduce the JS engine's force math.
//
// `Fixtures/conformance-golden.json` is emitted by `packages/core/src/conformance/golden-emit.ts`:
// the canonical deterministic forces fired at a fan of probe particles, each with its frame-0
// velocity delta (`dv`) as computed by the f64 JS engine. Here the f32 Swift engine applies the
// same force to the same reconstructed inputs and must land on the same `dv`.
//
// Tolerance is a single force apply (no integration accumulation), so f32↔f64 drift is tiny: an
// absolute 2e-4 with a small relative term covers it while still catching a real formula divergence
// (a wrong coefficient, a missing falloff leg, a sign flip). When this fails, the JS and Swift ports
// of that force have diverged — fix the Swift force, never loosen the tolerance to hide it.

private struct GoldenFile: Decodable {
    let count: Int
    let cases: [GoldenCase]
}

private struct GoldenCase: Decodable {
    struct BodyJSON: Decodable { let strength, range, spin, ux, uy: Float; let on: Bool }
    struct EnvJSON: Decodable { let dx, dy, dz, dist, orbit: Float }
    struct Vel: Decodable { let vx, vy, vz: Float }
    struct DV: Decodable { let x, y, z: Float }
    let force: String
    let label: String
    let px, py: Float
    let body: BodyJSON
    let env: EnvJSON
    let particle: Vel
    let dv: DV
}

@Suite("CrossPlaneConformance")
struct CrossPlaneConformanceTests {

    private func loadGolden() throws -> GoldenFile {
        let url = try #require(
            Bundle.module.url(forResource: "conformance-golden", withExtension: "json", subdirectory: "Fixtures"),
            "conformance-golden.json missing — run `node packages/core/src/conformance/golden-emit.ts`"
        )
        return try JSONDecoder().decode(GoldenFile.self, from: Data(contentsOf: url))
    }

    @Test("every golden case is reproduced by the Swift force within tolerance")
    func parity() throws {
        let golden = try loadGolden()
        #expect(golden.cases.count == golden.count && golden.count > 0)

        let registry = Registry.standard().forces
        var checked = 0

        for c in golden.cases {
            let force = try #require(registry[c.force], "Swift registry has no force '\(c.force)'")

            // reconstruct the exact inputs the JS apply saw.
            let body = Body(
                tokens: [c.force],
                strength: c.body.strength,
                range: c.body.range,
                spin: c.body.spin,
                heading: Vec3(c.body.ux, c.body.uy, 0),
                box: Box(center: .zero, halfExtents: .zero) // body at the origin (matches the emitter)
            )
            body.isEngaged = c.body.on

            let p = Particle(position: Vec3(c.px, c.py, 0),
                             velocity: Vec3(c.particle.vx, c.particle.vy, c.particle.vz))

            let env = Env()
            env.vector = Vec3(c.env.dx, c.env.dy, c.env.dz) // JS e.dx/dy/dz == Swift e.vector
            env.dist = c.env.dist
            env.form = Formation(orbit: c.env.orbit)

            let v0 = p.velocity
            force.apply(body: body, particle: p, env: env)
            let dv = p.velocity - v0
            let want = Vec3(c.dv.x, c.dv.y, c.dv.z)

            let tol: Float = 2e-4 + 1e-3 * simd_length(want)
            let err = simd_length(dv - want)
            #expect(
                err <= tol,
                "\(c.force)/\(c.label) @(\(c.px),\(c.py)): Swift dv=(\(dv.x),\(dv.y),\(dv.z)) vs JS (\(want.x),\(want.y),\(want.z)) — err \(err) > tol \(tol)"
            )
            checked += 1
        }
        #expect(checked == golden.count)
    }
}

File: android/fundamental-core/src/test/kotlin/com/fundamental/core/GoldenConformanceTests.kt
package com.fundamental.core

import com.fundamental.core.engine.Body
import com.fundamental.core.engine.Env
import com.fundamental.core.engine.Formation
import com.fundamental.core.engine.Particle
import com.fundamental.core.engine.Registry
import com.fundamental.core.math.Vec3
import kotlinx.serialization.Serializable
import kotlinx.serialization.json.Json
import kotlin.test.Test
import kotlin.test.assertEquals
import kotlin.test.assertTrue

// Cross-plane conformance (#526) — the Kotlin engine must reproduce the JS engine's force math.
//
// `fixtures/conformance-golden.json` (synced onto the classpath from the canonical Swift fixture by
// the `syncGolden` Gradle task — single source of truth) is emitted by
// `scripts/gen-conformance-golden.mjs`: the canonical deterministic forces fired at a fan of probe
// particles, each with its frame-0 velocity delta (`dv`) as computed by the f64 JS engine. Here the
// f32 Kotlin engine applies the same force to the same reconstructed inputs and must land on `dv`.
//
// Tolerance is a single force apply (no integration accumulation), so f32↔f64 drift is tiny: an
// absolute 2e-4 with a small relative term covers it while still catching a real formula divergence
// (a wrong coefficient, a missing falloff leg, a sign flip). When this fails, the JS and Kotlin
// ports of that force have diverged — fix the Kotlin force, never loosen the tolerance to hide it.

@Serializable
private data class GoldenFile(val count: Int, val cases: List<GoldenCase>)

@Serializable
private data class GoldenCase(
    val force: String,
    val label: String,
    val px: Float,
    val py: Float,
    val body: BodyJson,
    val env: EnvJson,
    val particle: Vel,
    val dv: DV,
)

@Serializable
private data class BodyJson(
    val strength: Float,
    val range: Float,
    val spin: Float,
    val on: Boolean,
    val ux: Float,
    val uy: Float,
)

@Serializable
private data class EnvJson(val dx: Float, val dy: Float, val dz: Float, val dist: Float, val orbit: Float)

@Serializable
private data class Vel(val vx: Float, val vy: Float, val vz: Float)

@Serializable
private data class DV(val x: Float, val y: Float, val z: Float)

class GoldenConformanceTests {

    private val json = Json { ignoreUnknownKeys = true }

    private fun loadGolden(): GoldenFile {
        val stream = javaClass.getResourceAsStream("/fixtures/conformance-golden.json")
            ?: error("conformance-golden.json missing on the classpath — run `pnpm gen:golden` (Gradle's syncGolden task pulls it in)")
        val text = stream.bufferedReader().use { it.readText() }
        return json.decodeFromString(GoldenFile.serializer(), text)
    }

    @Test
    fun parity() {
        val golden = loadGolden()
        assertTrue(golden.cases.size == golden.count && golden.count > 0, "golden file is empty or mis-counted")

        val registry = Registry.standardForces()
        var checked = 0

        for (c in golden.cases) {
            val force = registry[c.force] ?: error("Kotlin registry has no force '${c.force}'")

            // reconstruct the exact inputs the JS apply saw (matches GoldenConformanceTests.swift).
            val body = Body(
                tokens = listOf(c.force),
                strength = c.body.strength,
                range = c.body.range,
                spin = c.body.spin,
                heading = Vec3(c.body.ux, c.body.uy, 0f),
            )
            body.isEngaged = c.body.on

            val p = Particle(
                position = Vec3(c.px, c.py, 0f),
                velocity = Vec3(c.particle.vx, c.particle.vy, c.particle.vz),
            )

            val env = Env().apply {
                vector = Vec3(c.env.dx, c.env.dy, c.env.dz) // JS e.dx/dy/dz == Kotlin env.vector
                dist = c.env.dist
                form = Formation(orbit = c.env.orbit)
            }

            val v0 = p.velocity
            force.apply(body, p, env)
            val dv = p.velocity - v0
            val want = Vec3(c.dv.x, c.dv.y, c.dv.z)

            val tol = 2e-4f + 1e-3f * want.length()
            val err = (dv - want).length()
            assertTrue(
                err <= tol,
                "${c.force}/${c.label} @(${c.px},${c.py}): Kotlin dv=(${dv.x},${dv.y},${dv.z}) " +
                    "vs JS (${want.x},${want.y},${want.z}) — err $err > tol $tol",
            )
            checked++
        }
        assertEquals(golden.count, checked)
    }
}
